src/test/java/org/kiwiproject/registry/consul/server/ConsulRegistryServiceIntegrationTest.java

```diff
@@ -0,0 +1,199 @@
+            softly.assertThat(serviceInstance.getVersion()).isEqualTo(serviceInfo.getVersion());
+            softly.assertThat(serviceInstance.getDescription()).isEqualTo(serviceInfo.getDescription());
+            softly.assertThat(serviceInstance.getCommitRef()).isEqualTo(serviceInfo.getCommitRef());
+            softly.assertThat(serviceInstance.getPaths()).isEqualTo(serviceInfo.getPaths());
+            softly.assertThat(serviceInstance.getIp()).isEqualTo(serviceInfo.getIp());
+            softly.assertThat(serviceInstance.getStatus()).isEqualTo(ServiceInstance.Status.UP);
+        }
+
+    }
+
+    @Nested
+    class Register {
+
+        @Test
+        void shouldThrowIllegalStateException() {
+            service.registeredService.set(ServiceInstance.builder().build());
+
+            assertThatThrownBy(() -> service.register(ServiceInstance.builder().build()))
+                    .isInstanceOf(IllegalStateException.class)
+                    .hasMessageContaining("Cannot register. Already managing a registered instance");
+        }
+
+        @Test
+        void shouldRegisterAndReturnRegisteredInstance() {
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceInfo())
+                    .withStatus(ServiceInstance.Status.UP);
+
+            var registeredInstance = service.register(serviceInstance);
+
+            assertThat(registeredInstance).isNotSameAs(serviceInstance);
+
+            var storedInstance = service.registeredService.get();
+            assertThat(storedInstance).isNotNull();
+            assertThat(storedInstance.getInstanceId()).isNotBlank();
+
+            assertThat(consul.catalogClient().getService(serviceInstance.getServiceName()).getResponse()).isNotEmpty();
+        }
+
+        @Test
+        void shouldRetryRegisterAndThrowExceptionIfAllTriesExpire() {
+            var badConsul = mock(Consul.class);
+            var serviceToFail = new ConsulRegistryService(badConsul, new ConsulRegistrationConfig(), environment);
+
+            doThrow(new RuntimeException("Oops")).when(badConsul).agentClient();
+
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceInfo())
+                    .withStatus(ServiceInstance.Status.UP);
+
+            assertThatThrownBy(() -> serviceToFail.register(serviceInstance))
+                    .isInstanceOf(RegistrationException.class)
+                    .hasMessageStartingWith("Unable to register service " + serviceInstance.getServiceName() + ", id ")
+                    .hasMessageEndingWith(" with Consul after 60 attempts");
+
+            assertThat(serviceToFail.registeredService.get()).isNull();
```

Comment:
Not totally sure but I thought AssertJ has assertions for the `AtomicXxx` classes...so that you don't need to call `get()`

Alaso, just checked and there is an `AtomicReferenceAssert` that contains `hasValue` and `doesNotHaveValue` methods, so alas there isn't a `hasNoValue` method...

Reply:
Yeah that is why I did that, I will it had a `hasNoValue` assertion